src/core/algorithms/nar/value_range.cpp

```diff
@@ -0,0 +1,109 @@
+#include "value_range.h"
+
+namespace model {
+
+StringValueRange::StringValueRange(model::TypedColumnData const& column) {
+    domain = std::vector<std::string>();
+    for (size_t row_index = 0; row_index < column.GetNumRows(); row_index++) {
+        std::byte const* value = column.GetValue(row_index);
+        std::string string_value = model::Type::GetValue<std::string>(value);
+        bool first_occurrence =
+                std::find(domain.begin(), domain.end(), string_value) == domain.end();
+        if (first_occurrence) {
+            domain.emplace_back(string_value);
+        }
+    }
+}
+
+std::string StringValueRange::ToString() const {
+    std::string result;
+    result += "[";
+    if (domain.size() > 0) {
+        result += domain[0];
+    }
+    for (size_t i = 1; i < domain.size(); i++) {
+        result += (", " + domain[i]);
+    }
+    result += "]";
+    return result;
+}
+
+std::string IntValueRange::ToString() const {
+    std::string result;
+    result += "[";
+    result += std::to_string(lower_bound);
+    result += " - ";
+    result += std::to_string(upper_bound);
+    result += "]";
+    return result;
+}
+
+IntValueRange::IntValueRange(model::TypedColumnData const& column) {
+    bool initialized = false;
+    for (size_t row_index = 0; row_index < column.GetNumRows(); row_index++) {
+        std::byte const* value = column.GetValue(row_index);
+        long int int_value = model::Type::GetValue<model::Int>(value);
```

Comment:
You assign value of type `model::Int` to `int_value` of type `long`. Then you assign it to `lower_bound` and `upper_bound`, both of type `model::Int`. Why can't `int_value` be of type `model::Int`?
Moreover, `model::Int` is not the same type as `long` -- it's alias for `int_64_t`, which is twice as long on some machines (see https://en.cppreference.com/w/cpp/language/types).
```suggestion
        model::Int int_value = model::Type::GetValue<model::Int>(value);
```
The same for `DoubleValueRange`